### ISSUE TITLE
[Question] question title

### ISSUE THREAD

**Issue by gilcoder** [DLR-RM/stable-baselines3#715]
**Important Note: We do not do technical support, nor consulting** and don't answer personal questions per email.
Please post your question on the [RL Discord](https://discord.com/invite/xhfNqQv), [Reddit](https://www.reddit.com/r/reinforcementlearning/) or [Stack Overflow](https://stackoverflow.com/) in that case.

### Question

I'm trying to implement sampling and training asynchronously using the SAC algorithm. I made the attempt shown in the code below. But I always get an error because there seems to be a confusion between training and assessment modes. The training mode (False or True) is configured in the policy. And this is shared between the train and collect_rollouts methods. Is it possible to do collect_rollouts asynchronously?

### Additional context
Reference code:

`_rollouts_ = queue.Queue()

def train_async(sac, max_steps):
	global _rollouts_
	iteraction = 0
	while True:
		if not _rollouts_.empty():
			rollout = _rollouts_.get()
			if rollout is not None:
				iteraction += 1
				gradient_steps = sac.gradient_steps if sac.gradient_steps >= 0 else rollout.episode_timesteps
				if gradient_steps > 0:
					sac.train(gradient_steps, sac.batch_size)
			else:
				print("Training ending")
				break
		else:
			print("waiting for rollouts....")
			time.sleep(1)

def rollouts_async(sac, max_steps,  callback, log_interval=None):
	steps  = 0
	global _rollouts_
	while True:
		rollout = sac.collect_rollouts(sac.env, callback, sac.train_freq, sac.replay_buffer, sac.action_noise, sac.learning_starts, log_interval)
		if rollout.continue_training is False:
			_rollouts_.put(None)
			break
		else:
			_rollouts_.put( rollout )
			steps += 1
			if steps >= max_steps:
				_rollouts_.put(None)
				callback.on_training_end()
				break

def learn_async(sac, total_timesteps = 1000000, callback=None, log_interval=None, tb_log_name="run", reset_num_timesteps=True):
	total_timesteps, callback = sac._setup_learn(total_timesteps, None, callback, 0, 0, None, reset_num_timesteps, tb_log_name)
	callback.on_training_start(locals(), globals())
	t1 = threading.Thread(target=rollouts_async, args=(sac, total_timesteps, callback, log_interval))
	t1.start()
	t2 = threading.Thread(target=train_async, args=(sac, total_timesteps))
	t2.start()
	t1.join()
	t2.join()
`

### Checklist

- [ ] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**) OK
- [ ] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**) OK

**Comment by araffin** [comment]
Hello,
you can find a working proof of concept here: https://github.com/DLR-RM/rl-baselines3-zoo/blob/87001ed8a40f817d46c950e283d1ca29e405ad71/utils/callbacks.py#L95

(it is not polished but it works)

**Comment by gilcoder** [comment]
> Hello, you can find a working proof of concept here: https://github.com/DLR-RM/rl-baselines3-zoo/blob/87001ed8a40f817d46c950e283d1ca29e405ad71/utils/callbacks.py#L95
> 
> (it is not polished but it works)

Thank you.

**Comment by araffin** [comment]
> I didn't know you could do that with callbacks.

yes, callbacks & wrappers are quite powerful...

> It would be interesting if it was documented. 

Well, I didn't have time to polish it properly, I usually don't advertise experimental features... but if you do polish it, we would be happy to receive a PR that shows it as an example in the doc ;)

> I'm already using it and it's exactly what I wanted.

Good to hear =)

**Comment by gilcoder** [comment]
Hello dears. Well, let's go. A long time ago, I tested treaded training and sampling with DQN using tensorflow, but I couldn't get convergence using threading. I thought it would be different now with stable baselines3. But I ran two experiments with the same settings: the first with threads and the second without . The first does not converge. The second converges. I think there are two bigger possibilities. Either Pytorch, like Tensorflow/Keras, doesn't handle gradient descent well when using threads, or stable-baselines callback calls have some bugs. I have no idea what it is. So, for now, I'm training without threads. It's slower, but at least it converges.

**Comment by araffin** [comment]
Hello,
I'm actively using the callback, the important thing to check is comparing training with the same number of gradient updates.
And also comparing how long does it take to do the gradient update vs how long does it take to collect data.

I just used the callback yesterday and it worked fine:
- live training: https://www.twitch.tv/videos/1252962277 
- final policy: https://www.twitch.tv/givethatrobotacookie/clip/AlluringSpoopyToadItsBoshyTime-yb3MjHVBXlyc_kPl

Please note that we do not do technical support (so unless you provide a minimal example to reproduce the issue without a custom env, we won't give further answers).

**Comment by araffin** [comment]
the callback is now available via the `rl_zoo3` package (`rl_zoo3.callbacks`), `VecNormalize` checkpoints are missing though.
